casper/src/main/scala/coop/rchain/casper/util/rholang/RuntimeManager.scala

```diff
@@ -74,6 +86,171 @@ class RuntimeManagerImpl[F[_]: Concurrent: Metrics: Span: Log](
+                        systemDeploy.extractResult(par) match {
+                          case Right(result) =>
+                            verifyReplaySuccess(result).map(_.asRight[SystemDeployError]).pure
+                          case Left(systemDeployError: SystemDeployError) =>
+                            verifyReplayFailure(systemDeployError) match {
+                              case Some(replayFailure) => replayFailure.asLeft.pure
+                              case None =>
+                                runtime.replaySpace
+                                  .revertToSoftCheckpoint(softCheckpoint)
+                                  .as(systemDeployError.asLeft.asRight)
+                            }
+                          case Left(systemDeployPlatformFailure: SystemDeployPlatformFailure) =>
+                            systemDeployPlatformFailure.raiseError
+                        }
+                      case Some((_, unexpectedResults)) =>
+                        UnexpectedResult(unexpectedResults.flatMap(_.pars)).raiseError
+                      case None =>
+                        ConsumeFailed.raiseError
+                    }
+                }
+              }
+        }
+      }
+    }
+  }
+
+  def playSystemDeploy[S <: SystemDeploy](stateHash: StateHash)(
+      systemDeploy: S
+  ): F[SystemDeployPlayResult[systemDeploy.Result]] =
+    withResetRuntimeLock(stateHash) { runtime =>
+      playSystemDeployInternal(runtime)(systemDeploy) >>= {
+        case (eventLog, Right(result)) =>
+          runtime.space.createCheckpoint().map(_.root.toByteString) >>= { finalStateHash =>
+            SystemDeployPlayResult.playSucceeded(finalStateHash, eventLog, result).pure
+          }
+        case (eventLog, Left(systemDeployError)) =>
+          SystemDeployPlayResult.playFailed(eventLog, systemDeployError).pure
+      }
+    }
+
+  import coop.rchain.rspace.trace.Log
+  private def playSystemDeployInternal[S <: SystemDeploy](runtime: Runtime[F])(
+      systemDeploy: S
+  ): F[(Log, Either[SystemDeployError, systemDeploy.Result])] =
+    runtime.space.createSoftCheckpoint() >>= { preDeploySoftCheckpoint =>
+      evaluateSystemSource(runtime)(systemDeploy, replay = false) >>= {
+        case EvaluateResult(_, interpreterErrors) =>
+          if (interpreterErrors.nonEmpty) UnexpectedSystemErrors(interpreterErrors).raiseError
+          else
+            consumeResult(runtime)(systemDeploy, replay = false) >>= {
+              case Some((_, Seq(ListParWithRandom(Seq(par), _)))) =>
+                systemDeploy.extractResult(par) match {
+                  case Right(result) =>
+                    runtime.space
+                      .createSoftCheckpoint()
+                      .map(
+                        postDeploySoftCheckpoint => (postDeploySoftCheckpoint.log, result.asRight)
+                      )
+                  case Left(systemDeployError: SystemDeployError) =>
```

Comment:
An example for this case would be a failed `chargeDeploy`, correct?
Why can we revert to `preDeploySoftCheckpoint`? Couldn't this cause problems with replayability?

Reply:
Yes  like `insufficient funds`. Hmm, we're facing some errors in replays. Maybe your comment is spot on. @JosDenmark could you please take a look?

Reply:
While waiting for your assessment I removed these doubtful reverts in 0b62efd50073afc1c958d6a5642b3042a7cc92be

Reply:
Timm is right. In a forthcoming PR we fix these issues, reverting only when an user error (`InterpreterError`) is thrown during `replayDeploy` or `processDeploy`.